File: backend/src/core/redis_queue.py
"""
Redis Queue System - Python Implementation
使用Redis和RQ作为消息队列，提供高性能的任务处理
适用于需要高并发处理的场景
"""

import os
import json
import logging
import redis
from rq import Queue, Worker, Connection, Job
from rq.exceptions import NoSuchJobError
from typing import Dict, List, Optional, Any, Callable, Union
from datetime import datetime, timedelta
import threading
import time

# Set up logging
logger = logging.getLogger(__name__)

class RedisQueue:
    """使用Redis作为消息队列"""
    
    def __init__(self):
        # Redis configuration
        self.redis_config = {
            'host': os.getenv('REDIS_HOST', 'localhost'),
            'port': int(os.getenv('REDIS_PORT', '6379')),
            'db': int(os.getenv('REDIS_DB', '0')),
            'password': os.getenv('REDIS_PASSWORD', None),
            'decode_responses': True
        }
        
        # Initialize Redis connection
        self.redis_client = None
        self.queues = {}
        self.workers = {}
        self._initialized = False
        
        logger.info("Redis Queue initialized")
    
    def initialize(self):
        """初始化Redis连接和队列"""
        if self._initialized:
            return
            
        try:
            # Create Redis connection
            self.redis_client = redis.Redis(**self.redis_config)
            
            # Test connection
            self.redis_client.ping()
            
            self._initialized = True
            logger.info("Redis Queue initialized successfully")
            
        except Exception as e:
            logger.error(f"Failed to initialize Redis Queue: {e}")
            raise
    
    def get_queue(self, topic: str) -> Queue:
        """获取或创建指定主题的队列"""
        if not self._initialized:
            self.initialize()
            
        if topic not in self.queues:
            self.queues[topic] = Queue(
                name=topic,
                connection=self.redis_client,
                default_timeout=300  # 5分钟超时
            )
            logger.info(f"Created queue for topic: {topic}")
            
        return self.queues[topic]
    
    def publish(self, topic: str, message: Dict[str, Any], 
                priority: int = 0, delay_seconds: int = 0) -> str:
        """发布消息到队列
        
        Args:
            topic: 消息主题
            message: 消息内容
            priority: 优先级 (暂时不支持，RQ按FIFO处理)
            delay_seconds: 延迟发送秒数
            
        Returns:
            str: 任务ID
        """
        if not self._initialized:
            self.initialize()
            
        queue = self.get_queue(topic)
        
        # 准备任务数据
        task_data = {
            'topic': topic,
            'payload': message,
            'priority': priority,
            'published_at': datetime.now().isoformat()
        }
        
        try:
            if delay_seconds > 0:
                # 延迟任务
                job = queue.enqueue_in(
                    timedelta(seconds=delay_seconds),
                    process_redis_task,
                    task_data,
                    job_timeout=300
                )
            else:
                # 立即任务
                job = queue.enqueue(
                    process_redis_task,
                    task_data,
                    job_timeout=300
                )
            
            logger.info(f"Published message to Redis queue {topic}: {job.id}")
            return job.id
            
        except Exception as e:
            logger.error(f"Error publishing message to Redis: {e}")
            raise
    
    def subscribe(self, topic: str, callback: Callable[[Dict[str, Any], Dict[str, Any]], bool],
                 worker_name: str = None, burst: bool = False):
        """订阅主题消息
        
        Args:
            topic: 消息主题
            callback: 消息处理回调函数
            worker_name: 工作器名称
            burst: 是否为突发模式（处理完现有任务后停止）
        """
        if not self._initialized:
            self.initialize()
            
        if worker_name is None:
            worker_name = f"{topic}_worker_{int(time.time())}"
        
        queue = self.get_queue(topic)
        
        # 注册回调函数到全局处理器
        register_callback(topic, callback)
        
        # 创建工作器
        worker = Worker(
            [queue],
            connection=self.redis_client,
            name=worker_name
        )
        
        self.workers[f"{topic}_{worker_name}"] = worker
        
        # 启动工作器（在后台线程中）
        def start_worker():
            logger.info(f"Starting Redis worker for topic {topic}")
            try:
                worker.work(burst=burst)
            except Exception as e:
                logger.error(f"Error in Redis worker {worker_name}: {e}")
        
        worker_thread = threading.Thread(target=start_worker, daemon=True)
        worker_thread.start()
        
        logger.info(f"Subscribed to Redis queue topic {topic} with worker {worker_name}")
        
        return worker_name
    
    def get_queue_stats(self, topic: str = None) -> Dict[str, Any]:
        """获取队列统计信息"""
        if not self._initialized:
            self.initialize()
        
        if topic:
            # 单个队列统计
            queue = self.get_queue(topic)
            
            # 获取队列长度
            pending_count = len(queue)
            
            # 获取失败任务数
            failed_queue = Queue('failed', connection=self.redis_client)
            failed_jobs = failed_queue.get_jobs()
            topic_failed_count = sum(1 for job in failed_jobs 
                                   if job.meta.get('topic') == topic)
            
            return {
                'topic': topic,
                'pending_jobs': pending_count,
                'failed_jobs': topic_failed_count,
                'workers': len([w for k, w in self.workers.items() if k.startswith(topic)]),
                'total_jobs': pending_count + topic_failed_count
            }
        else:
            # 所有队列统计
            stats = {}
            for topic_name, queue in self.queues.items():
                pending_count = len(queue)
                failed_queue = Queue('failed', connection=self.redis_client)
                failed_jobs = failed_queue.get_jobs()
                topic_failed_count = sum(1 for job in failed_jobs 
                                       if job.meta.get('topic') == topic_name)
                
                stats[topic_name] = {
                    'pending_jobs': pending_count,
                    'failed_jobs': topic_failed_count,
                    'workers': len([w for k, w in self.workers.items() if k.startswith(topic_name)]),
                    'total_jobs': pending_count + topic_failed_count
                }
            
            return stats
    
    def get_job_status(self, job_id: str) -> Dict[str, Any]:
        """获取任务状态"""
        if not self._initialized:
            self.initialize()
            
        try:
            job = Job.fetch(job_id, connection=self.redis_client)
            
            return {
                'job_id': job_id,
                'status': job.get_status(),
                'created_at': job.created_at.isoformat() if job.created_at else None,
                'started_at': job.started_at.isoformat() if job.started_at else None,
                'ended_at': job.ended_at.isoformat() if job.ended_at else None,
                'result': job.result,
                'exc_info': job.exc_info,
                'meta': job.meta
            }
            
        except NoSuchJobError:
            return {'job_id': job_id, 'status': 'not_found'}
        except Exception as e:
            logger.error(f"Error getting job status: {e}")
            return {'job_id': job_id, 'status': 'error', 'error': str(e)}
    
    def cancel_job(self, job_id: str) -> bool:
        """取消任务"""
        if not self._initialized:
            self.initialize()
            
        try:
            job = Job.fetch(job_id, connection=self.redis_client)
            job.cancel()
            logger.info(f"Cancelled job: {job_id}")
            return True
            
        except NoSuchJobError:
            logger.warning(f"Job not found for cancellation: {job_id}")
            return False
        except Exception as e:
            logger.error(f"Error cancelling job {job_id}: {e}")
            return False
    
    def retry_failed_jobs(self, topic: str = None, limit: int = None) -> int:
        """重试失败的任务"""
        if not self._initialized:
            self.initialize()
            
        failed_queue = Queue('failed', connection=self.redis_client)
        failed_jobs = failed_queue.get_jobs()
        
        if topic:
            failed_jobs = [job for job in failed_jobs if job.meta.get('topic') == topic]
        
        if limit:
            failed_jobs = failed_jobs[:limit]
        
        retried_count = 0
        for job in failed_jobs:
            try:
                job.retry()
                retried_count += 1
                logger.info(f"Retried failed job: {job.id}")
            except Exception as e:
                logger.error(f"Error retrying job {job.id}: {e}")
        
        logger.info(f"Retried {retried_count} failed jobs")
        return retried_count
    
    def cleanup_old_jobs(self, days: int = 7) -> int:
        """清理旧任务"""
        if not self._initialized:
            self.initialize()
            
        cutoff_time = datetime.now() - timedelta(days=days)
        cleanup_count = 0
        
        # 清理已完成的任务
        for queue_name in self.queues.keys():
            queue = self.queues[queue_name]
            
            # 获取已完成的任务
            finished_jobs = queue.get_finished_job_registry()
            
            for job_id in finished_jobs.get_job_ids():
                try:
                    job = Job.fetch(job_id, connection=self.redis_client)
                    if job.ended_at and job.ended_at < cutoff_time:
                        job.delete()
                        cleanup_count += 1
                except Exception as e:
                    logger.error(f"Error cleaning up job {job_id}: {e}")
        
        logger.info(f"Cleaned up {cleanup_count} old jobs")
        return cleanup_count
    
    def stop_workers(self, topic: str = None):
        """停止工作器"""
        workers_to_stop = []
        
        if topic:
            workers_to_stop = [w for k, w in self.workers.items() if k.startswith(topic)]
        else:
            workers_to_stop = list(self.workers.values())
        
        for worker in workers_to_stop:
            try:
                worker.request_stop()
                logger.info(f"Requested stop for worker: {worker.name}")
            except Exception as e:
                logger.error(f"Error stopping worker {worker.name}: {e}")
    
    def close(self):
        """关闭Redis连接"""
        # 停止所有工作器
        self.stop_workers()
        
        # 清理资源
        self.queues.clear()
        self.workers.clear()
        
        if self.redis_client:
            self.redis_client.close()
            self.redis_client = None
        
        self._initialized = False
        logger.info("Redis Queue connections closed")

# 全局回调函数注册表
_callbacks = {}

def register_callback(topic: str, callback: Callable):
    """注册回调函数"""
    _callbacks[topic] = callback

def process_redis_task(task_data: Dict[str, Any]) -> bool:
    """处理Redis任务的通用函数"""
    topic = task_data.get('topic')
    payload = task_data.get('payload')
    
    if topic not in _callbacks:
        logger.error(f"No callback registered for topic: {topic}")
        return False
    
    try:
        callback = _callbacks[topic]
        
        # 准备元数据
        metadata = {
            'topic': topic,
            'published_at': task_data.get('published_at'),
            'started_at': datetime.now().isoformat()
        }
        
        # 调用回调函数
        result = callback(payload, metadata)
        
        logger.info(f"Successfully processed Redis task for topic: {topic}")
        return result if result is not None else True
        
    except Exception as e:
        logger.error(f"Error processing Redis task for topic {topic}: {e}")
        raise  # RQ会处理异常并标记任务为失败

# 全局Redis队列实例
redis_queue = RedisQueue()

File: backend/src/core/redisQueue.js
/**
 * Redis Queue System - JavaScript Implementation
 * 使用Redis和Bull作为消息队列，提供高性能的任务处理
 * 适用于需要高并发处理的场景
 */

const Bull = require('bull');
const Redis = require('ioredis');
const EventEmitter = require('events');

class RedisQueue extends EventEmitter {
  constructor() {
    super();
    
    // Redis configuration
    this.redisConfig = {
      host: process.env.REDIS_HOST || 'localhost',
      port: parseInt(process.env.REDIS_PORT || '6379'),
      db: parseInt(process.env.REDIS_DB || '0'),
      password: process.env.REDIS_PASSWORD || undefined,
      maxRetriesPerRequest: 3,
      retryDelayOnFailover: 100,
      lazyConnect: true
    };
    
    this.redis = null;
    this.queues = new Map();
    this.processors = new Map();
    this.initialized = false;
    
    console.log('Redis Queue initialized');
  }
  
  /**
   * 初始化Redis连接和队列系统
   */
  async initialize() {
    if (this.initialized) return;
    
    try {
      // Create Redis connection
      this.redis = new Redis(this.redisConfig);
      
      // Test connection
      await this.redis.ping();
      
      this.initialized = true;
      console.log('Redis Queue initialized successfully');
      
    } catch (error) {
      console.error('Failed to initialize Redis Queue:', error);
      throw error;
    }
  }
  
  /**
   * 获取或创建指定主题的队列
   */
  getQueue(topic) {
    if (!this.initialized) {
      throw new Error('Redis Queue not initialized');
    }
    
    if (!this.queues.has(topic)) {
      const queue = new Bull(topic, {
        redis: this.redisConfig,
        defaultJobOptions: {
          removeOnComplete: 50,    // 保留最近50个完成的任务
          removeOnFail: 100,       // 保留最近100个失败的任务
          attempts: 3,             // 最多重试3次
          backoff: {
            type: 'exponential',
            delay: 2000            // 指数退避，从2秒开始
          }
        }
      });
      
      this.queues.set(topic, queue);
      console.log(`Created Redis queue for topic: ${topic}`);
      
      // 监听队列事件
      queue.on('completed', (job, result) => {
        console.log(`Job ${job.id} completed in queue ${topic}`);
        this.emit('jobCompleted', { topic, jobId: job.id, result });
      });
      
      queue.on('failed', (job, err) => {
        console.error(`Job ${job.id} failed in queue ${topic}:`, err.message);
        this.emit('jobFailed', { topic, jobId: job.id, error: err.message });
      });
      
      queue.on('stalled', (job) => {
        console.warn(`Job ${job.id} stalled in queue ${topic}`);
        this.emit('jobStalled', { topic, jobId: job.id });
      });
    }
    
    return this.queues.get(topic);
  }
  
  /**
   * 发布消息到队列
   * @param {string} topic - 消息主题
   * @param {object} message - 消息内容
   * @param {object} options - 选项
   * @returns {Promise<string>} 任务ID
   */
  async publish(topic, message, options = {}) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const {
      priority = 0,
      delaySeconds = 0,
      maxRetries = 3
    } = options;
    
    const queue = this.getQueue(topic);
    
    // 准备任务数据
    const taskData = {
      topic,
      payload: message,
      publishedAt: new Date().toISOString()
    };
    
    // 设置任务选项
    const jobOptions = {
      priority: priority,
      attempts: maxRetries,
      backoff: {
        type: 'exponential',
        delay: 2000
      }
    };
    
    // 如果有延迟，设置延迟时间
    if (delaySeconds > 0) {
      jobOptions.delay = delaySeconds * 1000;
    }
    
    try {
      const job = await queue.add('process', taskData, jobOptions);
      
      console.log(`Published message to Redis queue ${topic}: ${job.id}`);
      this.emit('messagePublished', { topic, jobId: job.id, message });
      
      return job.id;
      
    } catch (error) {
      console.error('Error publishing message to Redis:', error);
      throw error;
    }
  }
  
  /**
   * 订阅主题消息
   * @param {string} topic - 消息主题
   * @param {function} callback - 消息处理回调函数
   * @param {object} options - 选项
   */
  async subscribe(topic, callback, options = {}) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const {
      concurrency = 1,
      processorName = 'process'
    } = options;
    
    const queue = this.getQueue(topic);
    
    // 注册处理器
    queue.process(processorName, concurrency, async (job) => {
      const { payload } = job.data;
      
      console.log(`Processing job ${job.id} in queue ${topic}`);
      
      try {
        // 准备元数据
        const metadata = {
          jobId: job.id,
          topic,
          attempt: job.attemptsMade + 1,
          maxAttempts: job.opts.attempts,
          publishedAt: job.data.publishedAt,
          startedAt: new Date().toISOString()
        };
        
        // 调用回调函数
        const result = await callback(payload, metadata);
        
        if (result === false) {
          throw new Error('Callback returned false, marking job as failed');
        }
        
        console.log(`Successfully processed job ${job.id} in queue ${topic}`);
        return result || true;
        
      } catch (error) {
        console.error(`Error processing job ${job.id} in queue ${topic}:`, error);
        throw error; // Bull会处理重试逻辑
      }
    });
    
    // 保存处理器信息
    this.processors.set(`${topic}_${processorName}`, {
      topic,
      callback,
      concurrency,
      processorName
    });
    
    console.log(`Subscribed to Redis queue topic ${topic} with concurrency ${concurrency}`);
    this.emit('subscribed', { topic, concurrency });
  }
  
  /**
   * 获取队列统计信息
   */
  async getQueueStats(topic = null) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    if (topic) {
      // 单个队列统计
      if (!this.queues.has(topic)) {
        return { topic, error: 'Queue not found' };
      }
      
      const queue = this.queues.get(topic);
      
      try {
        const [waiting, active, completed, failed, delayed] = await Promise.all([
          queue.getWaiting(),
          queue.getActive(),
          queue.getCompleted(),
          queue.getFailed(),
          queue.getDelayed()
        ]);
        
        return {
          topic,
          waiting: waiting.length,
          active: active.length,
          completed: completed.length,
          failed: failed.length,
          delayed: delayed.length,
          total: waiting.length + active.length + completed.length + failed.length + delayed.length
        };
        
      } catch (error) {
        console.error(`Error getting stats for queue ${topic}:`, error);
        return { topic, error: error.message };
      }
      
    } else {
      // 所有队列统计
      const stats = {};
      
      for (const [topicName, queue] of this.queues.entries()) {
        try {
          const [waiting, active, completed, failed, delayed] = await Promise.all([
            queue.getWaiting(),
            queue.getActive(),
            queue.getCompleted(),
            queue.getFailed(),
            queue.getDelayed()
          ]);
          
          stats[topicName] = {
            waiting: waiting.length,
            active: active.length,
            completed: completed.length,
            failed: failed.length,
            delayed: delayed.length,
            total: waiting.length + active.length + completed.length + failed.length + delayed.length
          };
          
        } catch (error) {
          console.error(`Error getting stats for queue ${topicName}:`, error);
          stats[topicName] = { error: error.message };
        }
      }
      
      return stats;
    }
  }
  
  /**
   * 获取任务状态
   */
  async getJobStatus(topic, jobId) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const queue = this.getQueue(topic);
    
    try {
      const job = await queue.getJob(jobId);
      
      if (!job) {
        return { jobId, status: 'not_found' };
      }
      
      return {
        jobId,
        status: await job.getState(),
        progress: job.progress(),
        createdAt: new Date(job.timestamp).toISOString(),
        processedOn: job.processedOn ? new Date(job.processedOn).toISOString() : null,
        finishedOn: job.finishedOn ? new Date(job.finishedOn).toISOString() : null,
        attemptsMade: job.attemptsMade,
        maxAttempts: job.opts.attempts,
        data: job.data,
        returnvalue: job.returnvalue,
        failedReason: job.failedReason
      };
      
    } catch (error) {
      console.error(`Error getting job status: ${error}`);
      return { jobId, status: 'error', error: error.message };
    }
  }
  
  /**
   * 取消任务
   */
  async cancelJob(topic, jobId) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const queue = this.getQueue(topic);
    
    try {
      const job = await queue.getJob(jobId);
      
      if (!job) {
        console.warn(`Job not found for cancellation: ${jobId}`);
        return false;
      }
      
      await job.remove();
      console.log(`Cancelled job: ${jobId}`);
      this.emit('jobCancelled', { topic, jobId });
      
      return true;
      
    } catch (error) {
      console.error(`Error cancelling job ${jobId}:`, error);
      return false;
    }
  }
  
  /**
   * 重试失败的任务
   */
  async retryFailedJobs(topic, limit = null) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const queue = this.getQueue(topic);
    
    try {
      let failedJobs = await queue.getFailed();
      
      if (limit) {
        failedJobs = failedJobs.slice(0, limit);
      }
      
      let retriedCount = 0;
      
      for (const job of failedJobs) {
        try {
          await job.retry();
          retriedCount++;
          console.log(`Retried failed job: ${job.id}`);
        } catch (error) {
          console.error(`Error retrying job ${job.id}:`, error);
        }
      }
      
      console.log(`Retried ${retriedCount} failed jobs in queue ${topic}`);
      this.emit('jobsRetried', { topic, count: retriedCount });
      
      return retriedCount;
      
    } catch (error) {
      console.error(`Error retrying failed jobs:`, error);
      return 0;
    }
  }
  
  /**
   * 清理旧任务
   */
  async cleanupOldJobs(topic = null, days = 7) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const cutoffTime = Date.now() - (days * 24 * 60 * 60 * 1000);
    let totalCleaned = 0;
    
    const queuesToClean = topic ? [this.getQueue(topic)] : Array.from(this.queues.values());
    
    for (const queue of queuesToClean) {
      try {
        // 清理已完成的任务
        const completed = await queue.getCompleted();
        let cleanedFromQueue = 0;
        
        for (const job of completed) {
          if (job.finishedOn < cutoffTime) {
            await job.remove();
            cleanedFromQueue++;
          }
        }
        
        // 清理失败的任务
        const failed = await queue.getFailed();
        
        for (const job of failed) {
          if (job.finishedOn < cutoffTime) {
            await job.remove();
            cleanedFromQueue++;
          }
        }
        
        totalCleaned += cleanedFromQueue;
        console.log(`Cleaned ${cleanedFromQueue} old jobs from queue ${queue.name}`);
        
      } catch (error) {
        console.error(`Error cleaning queue ${queue.name}:`, error);
      }
    }
    
    console.log(`Total cleaned ${totalCleaned} old jobs`);
    this.emit('jobsCleaned', { count: totalCleaned, days });
    
    return totalCleaned;
  }
  
  /**
   * 暂停队列
   */
  async pauseQueue(topic) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const queue = this.getQueue(topic);
    await queue.pause();
    
    console.log(`Paused queue: ${topic}`);
    this.emit('queuePaused', { topic });
  }
  
  /**
   * 恢复队列
   */
  async resumeQueue(topic) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const queue = this.getQueue(topic);
    await queue.resume();
    
    console.log(`Resumed queue: ${topic}`);
    this.emit('queueResumed', { topic });
  }
  
  /**
   * 关闭队列连接
   */
  async close() {
    console.log('Closing Redis Queue connections...');
    
    // 关闭所有队列
    for (const [topic, queue] of this.queues.entries()) {
      try {
        await queue.close();
        console.log(`Closed queue: ${topic}`);
      } catch (error) {
        console.error(`Error closing queue ${topic}:`, error);
      }
    }
    
    // 关闭Redis连接
    if (this.redis) {
      try {
        await this.redis.quit();
        this.redis = null;
      } catch (error) {
        console.error('Error closing Redis connection:', error);
      }
    }
    
    // 清理资源
    this.queues.clear();
    this.processors.clear();
    this.initialized = false;
    
    console.log('Redis Queue connections closed');
    this.emit('closed');
  }
}

module.exports = RedisQueue;